File: harvester/src/harvester/stats.py
"""Fetch-outcome scoreboard: an append-only JSONL of every dispatch result.

Every `get_or_fetch` terminal result appends one line to `<cache_root>/stats.jsonl`:
``{"ts": ..., "item": ..., "ok": true|false, "detail": "method-or-error_kind"}``. Aggregated,
this answers the questions the per-artifact `rungs:` frontmatter cannot: WHICH sources win,
which failure kinds dominate, and what a source's real success rate is over time — so a dying
OA source or a newly walled publisher is VISIBLE instead of silently rotting.

Nothing here raises or blocks the fetch path: a failed append is logged and dropped. Run
`python -m harvester.stats` for the human-readable summary table.
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

from .cache import cache_root
from .log import get_logger

log = get_logger("stats")

STATS_FILENAME = "stats.jsonl"
_MAX_ITEM_LEN = 500
_MAX_DETAIL_LEN = 200


def stats_path() -> Path:
    """The JSONL file lives INSIDE the cache root, so it follows WEBFETCH_DIR/HARVESTER_CACHE_DIR."""
    return cache_root() / STATS_FILENAME


def record_fetch(item: str, ok: bool, detail: str = "") -> None:
    """Append one outcome line. Never raises — telemetry must never break a fetch."""
    entry = {
        "ts": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "item": (item or "")[:_MAX_ITEM_LEN],
        "ok": bool(ok),
        "detail": (detail or "")[:_MAX_DETAIL_LEN],
    }
    try:
        with open(stats_path(), "a", encoding="utf-8") as fh:
            fh.write(json.dumps(entry, ensure_ascii=False) + "\n")
    except OSError as e:
        log.warning("stats append failed: %s", e)


def summarize(last_n: int = 5000) -> dict[str, dict[str, float | int]]:
    """Aggregate the last *last_n* records by `detail` → {total, ok, rate}.

    Reads at most the tail of the file (a scoreboard that loads forever would defeat itself).
    Returns {} when nothing has been recorded yet — an EMPTY result means no data was ever
    written, which is distinguishable from a broken reader by the caller checking that
    `stats_path()` exists.
    """
    path = stats_path()
    try:
        size = path.stat().st_size
    except OSError as e:
        log.warning("stats summarize could not stat %s: %s", path, e)
        return {}
    try:
        with open(path, "rb") as fh:
            if size > last_n * 512:  # seek near the end; line boundaries re-split below
                fh.seek(max(0, size - last_n * 512))
                fh.readline()  # drop the (possibly partial) first line
            raw = fh.read()
    except OSError as e:
        log.warning("stats summarize could not read %s: %s", path, e)
        return {}
    buckets: dict[str, dict[str, float | int]] = {}
    seen = 0
    for line in raw.decode("utf-8", errors="ignore").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            rec = json.loads(line)
            ok = bool(rec.get("ok"))
            detail = str(rec.get("detail") or ("ok" if ok else "unknown"))
        except (json.JSONDecodeError, AttributeError) as e:
            log.debug("stats skipping malformed line %r: %s", line[:80], e)
            continue
        seen += 1
        b = buckets.setdefault(detail, {"total": 0, "ok": 0})
        b["total"] = int(b["total"]) + 1
        if ok:
            b["ok"] = int(b["ok"]) + 1
    for b in buckets.values():
        total = int(b["total"])
        b["rate"] = round(int(b["ok"]) / total, 3) if total else 0.0
    log.info("stats summarized %d record(s) -> %d detail bucket(s)", seen, len(buckets))
    return buckets


def main() -> None:  # pragma: no cover — thin CLI wrapper over summarize()
    buckets = summarize()
    path = stats_path()
    if not buckets:
        if not path.exists():
            print(f"No stats recorded yet ({path} does not exist). Fetch something first.")
        else:
            print(f"Stats file {path} exists but holds no parseable records.")
        raise SystemExit(1)
    width = max(len(k) for k in buckets)
    print(f"{'detail'.ljust(width)}  {'ok':>6}  {'total':>6}  {'rate':>6}")
    for k in sorted(buckets, key=lambda k: -int(buckets[k]['total'])):
        b = buckets[k]
        print(f"{k.ljust(width)}  {b['ok']:>6}  {b['total']:>6}  {b['rate']:>6.3f}")


if __name__ == "__main__":
    main()

File: pfm/internal/harvest/stats.go
package harvest

// Fetch-outcome scoreboard (parity with harvester stats.py): an append-only
// JSONL of every terminal dispatch outcome under <CacheDir>/stats.jsonl.
// Aggregated, it answers what the per-artifact `rungs:` trace cannot: WHICH
// sources win, which failure kinds dominate, and a source's real success rate
// over time — so a dying OA provider or a newly walled publisher is VISIBLE
// instead of silently rotting.
//
// Nothing here raises into the fetch path: a failed append is logged and dropped.

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const statsFilename = "stats.jsonl"

type statRecord struct {
	TS     string `json:"ts"`
	Item   string `json:"item"`
	OK     bool   `json:"ok"`
	Detail string `json:"detail"`
}

func (h *Harvester) recordStat(item string, result Result) {
	dir := h.options.CacheDir
	if dir == "" {
		return
	}
	rec := statRecord{
		TS:   time.Now().UTC().Format("2006-01-02T15:04:05Z"),
		Item: truncateRunes(item, 500),
		OK:   result.Error == "",
	}
	if rec.OK {
		rec.Detail = truncateRunes(result.Method, 200)
	} else if result.ErrorKind != "" {
		rec.Detail = truncateRunes(result.ErrorKind, 200)
	} else {
		rec.Detail = "error"
	}
	line, err := json.Marshal(rec)
	if err != nil {
		log.Printf("harvest: stats marshal failed: %v", err)
		return
	}
	if err := os.MkdirAll(dir, 0o700); err != nil {
		log.Printf("harvest: create stats directory %s: %v", dir, err)
		return
	}
	file, err := os.OpenFile(filepath.Join(dir, statsFilename), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		log.Printf("harvest: stats append failed: %v", err)
		return
	}
	defer func() {
		if closeErr := file.Close(); closeErr != nil {
			log.Printf("harvest: stats close failed: %v", closeErr)
		}
	}()
	if _, err := file.Write(append(line, '\n')); err != nil {
		log.Printf("harvest: stats write failed: %v", err)
	}
}

// StatBucket aggregates one detail's outcomes.
type StatBucket struct {
	Total int     `json:"total"`
	OK    int     `json:"ok"`
	Rate  float64 `json:"rate"`
}

// SummarizeStats reads the scoreboard tail from *cacheDir* and buckets the last
// records by detail → {total, ok, rate}. An empty map with no error means the
// file does not exist or holds no parseable records — distinguishable at the
// caller by checking the file first.
func SummarizeStats(cacheDir string, lastN int) (map[string]*StatBucket, error) {
	path := filepath.Join(cacheDir, statsFilename)
	file, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]*StatBucket{}, nil
		}
		return nil, fmt.Errorf("open stats file: %w", err)
	}
	defer file.Close()
	if lastN <= 0 {
		lastN = 5000
	}
	// Tail read: skip to near the end, drop the possibly partial line.
	info, err := file.Stat()
	if err != nil {
		return nil, fmt.Errorf("stat stats file: %w", err)
	}
	if size := info.Size(); size > int64(lastN)*512 {
		if _, err := file.Seek(size-int64(lastN)*512, 0); err == nil {
			reader := bufio.NewReader(file)
			_, _ = reader.ReadString('\n')
		} else {
			if _, seekErr := file.Seek(0, 0); seekErr != nil {
				return nil, fmt.Errorf("seek stats file: %w", seekErr)
			}
		}
	}
	buckets := map[string]*StatBucket{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var rec statRecord
		if err := json.Unmarshal([]byte(line), &rec); err != nil {
			continue // malformed lines are skipped, never fatal
		}
		detail := rec.Detail
		if detail == "" {
			if rec.OK {
				detail = "ok"
			} else {
				detail = "unknown"
			}
		}
		bucket, found := buckets[detail]
		if !found {
			bucket = &StatBucket{}
			buckets[detail] = bucket
		}
		bucket.Total++
		if rec.OK {
			bucket.OK++
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("scan stats file: %w", err)
	}
	for _, bucket := range buckets {
		if bucket.Total > 0 {
			bucket.Rate = float64(bucket.OK) / float64(bucket.Total)
		}
	}
	return buckets, nil
}
